api/src/test/java/org/openmrs/api/UserServiceTest.java

```diff
@@ -152,7 +156,107 @@ public void saveUser_shouldShouldCreateUserWhoIsPatientAlready() throws SQLExcep
 		assertEquals(1, allPatientsSet.size());
 	}
-	
+
+	@Test(expected = APIException.class)
+	public void createUser_shouldNotAllowExistingUser() {
+		UserService userService = Context.getUserService();
+		User someUser = getBrunoUser(userService);
+
+		userService.createUser(someUser, SOME_VALID_PASSWORD);
+	}
+
+	@Test(expected = APIException.class)
+	public void createUser_shouldNotAllowBlankPassword() {
+		UserService userService = Context.getUserService();
+		User unsavedUser = new User();
+
+		userService.createUser(unsavedUser, "");
+	}
+
+	@Test(expected = APIException.class)
+	public void createUser_shouldNotAllowNullPassword() {
+		UserService userService = Context.getUserService();
+		User unsavedUser = new User();
+
+		userService.createUser(unsavedUser, null);
+	}
+
+	@Test(expected = DAOException.class)
+	public void createUser_shouldNotAllowForDuplicatedUsername() {
+		UserService userService = Context.getUserService();
+		User someUser = getBrunoUser(userService);
+
+		User newUser = userWithValidPerson();
+		newUser.setUsername(someUser.getUsername());
+
+		userService.createUser(newUser, SOME_VALID_PASSWORD);
+	}
+
+	@Test(expected = DAOException.class)
+	public void createUser_shouldNotAllowDuplicatedSystemId() {
+		UserService userService = Context.getUserService();
+		User someUser = getBrunoUser(userService);
+
+		User newUser = userWithValidPerson();
+		newUser.setSystemId(someUser.getSystemId());
+
+		userService.createUser(newUser, SOME_VALID_PASSWORD);
+	}
+
+	@Test(expected = DAOException.class)
+	public void createUser_shouldNotAllowUsernameEqualsExistingSystemId() {
+		UserService userService = Context.getUserService();
+		User someUser = getBrunoUser(userService);
+
+		User newUser = userWithValidPerson();
+		newUser.setUsername(someUser.getSystemId());
+
+		userService.createUser(newUser, SOME_VALID_PASSWORD);
+	}
+
+	@Test(expected = DAOException.class)
+	public void createUser_shouldNotAllowSystemIdEqualsExistingUsername() {
+		UserService userService = Context.getUserService();
+		User someUser = getBrunoUser(userService);
+
+		User newUser = userWithValidPerson();
+		newUser.setSystemId(someUser.getUsername());
+
+		userService.createUser(newUser, SOME_VALID_PASSWORD);
+	}
+
+	@Test(expected = DAOException.class)
+	public void createUser_shouldNotAllowSystemIdEqualsUsernameWithLuhnCheckDigit() {
+		UserService userService = Context.getUserService();
+		User someUser = getBrunoUser(userService);
+
+		User newUser = userWithValidPerson();
+		newUser.setUsername(someUser.getUsername());
+		newUser.setSystemId(decorateWithLuhnIdentifier(someUser.getUsername()));
+
+		userService.createUser(newUser, SOME_VALID_PASSWORD);
+	}
+
+	private User userWithValidPerson() {
+		Person person = new Person();
+		person.addName(new PersonName("jane", "sue", "doe"));
+		person.setGender("F");
+		return new User(person);
+	}
+
+	private User getBrunoUser(UserService userService) {
```

Comment:
@wisogetnid can you replace this method with `userService.getUserByUsername("bruno")`?